File: CSharp/src/Runtime/Scripts/XEvent/manager.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XEvent
    {
        public delegate void Callback(params object[] args);

        public class Handler
        {
            public Callback Func;

            public Delegate Origin;

            public bool Once;
        }

        public partial class Manager : IDisposable
        {
            public bool Singleton { get; set; }

            public Dictionary<int, List<Handler>> Handlers { get; protected set; } = new();

            protected readonly ReaderWriterLockSlim Mutex = new();

            protected readonly ThreadLocal<List<Handler>> Batches = new(() => new List<Handler>(64));

            public void Dispose() { Mutex?.Dispose(); Batches?.Dispose(); }

            public virtual bool Register(Enum id, Callback callback, bool once = false) { return Register(id.GetHashCode(), callback, once); }

            public virtual bool Register(int id, Callback callback, bool once = false)
            {
                if (callback == null)
                {
                    XLog.Error("XEvent.Manager.Register: nil callback, id={0}.", id);
                    return false;
                }
                Mutex.EnterWriteLock();
                try
                {
                    if (Handlers.TryGetValue(id, out List<Handler> handlers) == false)
                    {
                        handlers = new List<Handler>();
                        Handlers.Add(id, handlers);
                    }
                    if (Singleton && handlers.Count > 0)
                    {
                        XLog.Error("XEvent.Manager.Register: singleton mode doesn't allow multiple registrations, id={0}.", id);
                        return false;
                    }
                    for (var i = 0; i < handlers.Count; i++)
                    {
                        if (Equals(handlers[i].Origin, callback)) return false;
                    }
                    var handler = new Handler { Func = callback, Origin = callback, Once = once };
                    handlers.Add(handler);
                    return true;
                }
                finally { Mutex.ExitWriteLock(); }
            }

            public virtual bool Unregister(Enum id, Callback callback = null) { return Unregister(id.GetHashCode(), callback); }

            public virtual bool Unregister(int id, Callback callback = null)
            {
                Mutex.EnterWriteLock();
                try
                {
                    var ret = false;
                    if (Handlers.TryGetValue(id, out var handlers))
                    {
                        if (callback != null)
                        {
                            if (handlers.Count > 0)
                            {
                                ret = handlers.RemoveAll(h => Equals(h.Origin, callback)) > 0;
                                if (handlers.Count == 0) Handlers.Remove(id);
                            }
                        }
                        else
                        {
                            ret = handlers.Count > 0;
                            Handlers.Remove(id);
                        }
                    }
                    return ret;
                }
                finally { Mutex.ExitWriteLock(); }
            }

            public virtual void Clear()
            {
                Mutex.EnterWriteLock();
                try { Handlers.Clear(); }
                finally { Mutex.ExitWriteLock(); }
            }

            public virtual bool Notify(Enum id, params object[] args) { return Notify(id.GetHashCode(), args); }

            public virtual bool Notify(int id, params object[] args)
            {
                var batches = Batches.Value;
                Mutex.EnterReadLock();
                try
                {
                    if (batches.Count > 0) batches.Clear();
                    if (Handlers.TryGetValue(id, out var handlers))
                    {
                        if (handlers != null && handlers.Count > 0)
                        {
                            for (int i = 0; i < handlers.Count; i++)
                            {
                                var handler = handlers[i];
                                if (handler != null && handler.Func != null)
                                {
                                    batches.Add(handler);
                                }
                            }
                        }
                    }
                }
                finally { Mutex.ExitReadLock(); }

                if (batches.Count > 0)
                {
                    var once = false;
                    for (int i = 0; i < batches.Count; i++)
                    {
                        var handler = batches[i];
                        if (handler.Once) once = true;
                        handler?.Func?.Invoke(args);
                    }
                    if (once)
                    {
                        Mutex.EnterWriteLock();
                        try
                        {
                            if (Handlers.TryGetValue(id, out var nhandlers))
                            {
                                nhandlers.RemoveAll(h => batches.Any(o => o.Once && object.ReferenceEquals(o.Origin, h.Origin)));
                                if (nhandlers.Count == 0) Handlers.Remove(id);
                            }
                        }
                        finally { Mutex.ExitWriteLock(); }
                    }
                    batches.Clear();
                    return true;
                }
                return false;
            }
        }
    }
}

File: Go/XEvent/manager.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XEvent

import (
	"reflect"
	"slices"
	"sync"

	"github.com/frameworkease/Universal.X.Utility/Go/XLog"
)

type Callback func(args ...any)

type Handler struct {
	Func   Callback
	Origin uintptr
	Once   bool
}

type Manager struct {
	Mutex    sync.RWMutex
	Singleton bool
	Handlers map[int][]*Handler
}

func (manager *Manager) Register(id int, callback Callback, once ...bool) bool {
	if nil == callback {
		XLog.Error("XEvent.Manager.Register: nil callback, id=%v", id)
		return false
	}
	manager.Mutex.Lock()
	defer manager.Mutex.Unlock()

	if manager.Handlers == nil {
		manager.Handlers = make(map[int][]*Handler)
	}

	handlers, ok := manager.Handlers[id]
	if !ok {
		handlers = make([]*Handler, 0)
	}

	if manager.Singleton && len(handlers) > 0 {
		XLog.Error("XEvent.Manager.Register: singleton mode doesn't allow multiple registrations, id=%v", id)
		return false
	}

	cptr := uintptr(reflect.ValueOf(callback).Pointer())
	for i := 0; i < len(handlers); i++ {
		if handlers[i].Origin == cptr {
			return false
		}
	}

	handler := new(Handler)
	handler.Func = callback
	handler.Origin = cptr
	handler.Once = len(once) > 0 && once[0]
	handlers = append(handlers, handler)
	manager.Handlers[id] = handlers

	return true
}

func (manager *Manager) Unregister(id int, callback ...Callback) bool {
	manager.Mutex.Lock()
	defer manager.Mutex.Unlock()

	if manager.Handlers == nil {
		return false
	}

	handlers, ok := manager.Handlers[id]
	if !ok {
		return false
	}

	sig := false
	var hptr uintptr = 0
	if len(callback) > 0 && callback[0] != nil {
		hptr = uintptr(reflect.ValueOf(callback[0]).Pointer())
	}
	if hptr != 0 {
		nhandlers := slices.DeleteFunc(handlers, func(ele *Handler) bool {
			ok := ele.Origin == hptr
			return ok
		})
		sig = len(nhandlers) != len(handlers)
		if len(nhandlers) == 0 {
			delete(manager.Handlers, id)
		} else {
			manager.Handlers[id] = nhandlers
		}
	} else {
		sig = len(handlers) > 0
		if sig {
			delete(manager.Handlers, id)
		}
	}
	return sig
}

func (manager *Manager) Clear() {
	manager.Mutex.Lock()
	defer manager.Mutex.Unlock()

	if manager.Handlers == nil {
		return
	} else {
		manager.Handlers = make(map[int][]*Handler)
	}
}

func (manager *Manager) Notify(id int, args ...any) bool {
	manager.Mutex.RLock()
	if manager.Handlers == nil {
		manager.Mutex.RUnlock()
		return false
	} else {
		handlers, ok := manager.Handlers[id]
		manager.Mutex.RUnlock()
		if !ok || len(handlers) == 0 {
			return false
		}

		once := false
		for _, handler := range handlers {
			if handler != nil && handler.Func != nil {
				if handler.Once {
					once = true
				}
				handler.Func(args...)
			}
		}

		if once {
			manager.Mutex.Lock()
			defer manager.Mutex.Unlock()

			nhandlers, ok := manager.Handlers[id]
			if ok {
				onces := make(map[uintptr]bool)
				for _, handler := range handlers {
					if handler.Once {
						onces[handler.Origin] = true
					}
				}
				nhandlers = slices.DeleteFunc(nhandlers, func(ele *Handler) bool { return onces[ele.Origin] })
				if len(nhandlers) == 0 {
					delete(manager.Handlers, id)
				} else {
					manager.Handlers[id] = nhandlers
				}
			}
		}

		return true
	}
}
